pkg/playground/speed.js

```diff
@@ -1,136 +1,179 @@
-                    ws.send(input);
+                    channel.send(input);
             }, interval);
-        };
+        }
+    }
 
-        ws.onmessage = function(event) {
-            if (binary)
-                total += event.data.byteLength;
-            else
-                total += event.data.length;
-        };
+    function read(ev) {
+        stop();
 
-        var update = window.setInterval(function() {
-            speed(total, start, "...");
-        }, 500);
+        var sideband = ev.target.id == "read-sideband";
+        var path = document.getElementById("read-path");
 
-        ws.onclose = function(event) {
-            if (!done)
-                window.alert("channel closed");
+        var options = {
+            payload: "fsread1",
+            path: path.value,
+            binary: sideband ? "raw" : true,
         };
 
-        window.setTimeout(function() {
-            done = true;
-            window.clearInterval(timer);
-            window.clearInterval(update);
+        start = Date.now();
+        total = 0;
+
+        if (sideband) {
+            websocket = new window.WebSocket(cockpit.transport.uri("channel/" + cockpit.transport.csrf_token) +
+                "?" + window.btoa(JSON.stringify(options)));
+            websocket.binaryType = 'arraybuffer';
+            websocket.onmessage = function(event) {
+                total += event.data.byteLength;
+            };
+            websocket.onclose = function(event) {
+                if (websocket)
+                    window.alert("channel closed");
+            };
+        } else {
+            channel = cockpit.channel(options);
+            channel.addEventListener("message", function(event, data) {
+                total += data.length;
+            });
+            channel.addEventListener("close", function(event, options) {
+                if (options.problem)
+                    window.alert(options.problem);
+            });
+        }
+    }
+
+    function stop() {
+        update();
+
+        if (channel)
+            channel.close();
+        channel = null;
+        var ws = websocket;
+        websocket = null;
+        if (ws)
             ws.close();
-            speed(total, start);
-        }, 10000);
+
+        window.clearInterval(timer);
+        timer = null;
     }
 
     cockpit.transport.wait(function() {
-        $("#normal").on("click", normal);
-        $("#sideband").on("click", sideband);
-        $("body").show();
+        document.getElementById("echo-normal").addEventListener("click", echo);
+        document.getElementById("echo-sideband").addEventListener("click", echo);
+        document.getElementById("read-normal").addEventListener("click", read);
+        document.getElementById("read-sideband").addEventListener("click", read);
+        document.getElementById("stop").addEventListener("click", stop);
+        window.setInterval(update, 500);
+        document.body.style.display = "block";
```

Comment:
That looks a bit odd - aren't we usually using `hidden` and `element.removeAttribute("hidden")`?
But this is total bikeshedding of course, not not at all important.

Reply:
Makes sense. I'll do that in the future. Let me know if i should do that here.

Reply:
No need to reset the tests and spend time on this IMHO; I just caught my eye, as it's a rather unusual way of showing an element.